hdf5 1.12.0 API changes

Hi! We're currently doing rebuilds on Arch Linux for hdf5 1.12.0.
[Upstream introduced breaking API changes in that release](https://portal.hdfgroup.org/display/HDF5/API+Compatibility+Reports+for+1.12).

This breaks the build for csound's hdf5 opcode:

```
[ 78%] Building C object Opcodes/hdf5/CMakeFiles/hdf5ops.dir/HDF5IO.c.o
cd /build/csound/src/csound-6.14.0/build/Opcodes/hdf5 && /usr/bin/cc -DCS_DEFAULT_PLUGINDIR=\"/usr/lib/csound/plugins64-6.0\" -DHAVE_EXECINFO -DHAVE_SOCKETS -DHAVE_STRTOD_L -DHAVE_STRTOK_R -DLINUX -DNO_FLTK_THREADS -DPIPES -DUSE_LRINT -D_CSOUND_RELEASE_ -D_GNU_SOURCE -Dhdf5ops_EXPORTS -I/build/csound/src/csound-6.14.0
/./H -I/build/csound/src/csound-6.14.0/./include -I/build/csound/src/csound-6.14.0/./Engine -I/build/csound/src/csound-6.14.0/./util -I/build/csound/src/csound-6.14.0/. -I/build/csound/src/csound-6.14.0/build -I/build/csound/src/csound-6.14.0/build/include  -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -ffast-math -
mfpmath=sse -msse2 -fomit-frame-pointer -DNDEBUG -fvisibility=hidden -std=gnu99 -DHAVE_ATOMIC_BUILTIN -DHAVE_CURL -DHAVE_PTHREAD -O3  -fPIC   -o CMakeFiles/hdf5ops.dir/HDF5IO.c.o   -c /build/csound/src/csound-6.14.0/Opcodes/hdf5/HDF5IO.c
In file included from /build/csound/src/csound-6.14.0/./include/csoundCore.h:35,
                 from /build/csound/src/csound-6.14.0/./include/csdl.h:111,
                 from /build/csound/src/csound-6.14.0/Opcodes/hdf5/HDF5IO.h:20,
                 from /build/csound/src/csound-6.14.0/Opcodes/hdf5/HDF5IO.c:20:
/build/csound/src/csound-6.14.0/Opcodes/hdf5/HDF5IO.c: In function ‘HDF5IO_readStringAttribute’:
/build/csound/src/csound-6.14.0/Opcodes/hdf5/HDF5IO.c:232:15: error: too few arguments to function ‘H5Oget_info3’
  232 |     HDF5ERROR(H5Oget_info(dataSetID, &oinfo));
      |               ^~~~~~~~~~~
/build/csound/src/csound-6.14.0/./include/sysdep.h:345:45: note: in definition of macro ‘UNLIKELY’
  345 | #  define UNLIKELY(x)   __builtin_expect(!!(x),0)
      |                                             ^
/build/csound/src/csound-6.14.0/Opcodes/hdf5/HDF5IO.c:232:5: note: in expansion of macro ‘HDF5ERROR’
  232 |     HDF5ERROR(H5Oget_info(dataSetID, &oinfo));
      |     ^~~~~~~~~
In file included from /usr/include/H5Apublic.h:22,
                 from /usr/include/hdf5.h:23,
                 from /build/csound/src/csound-6.14.0/Opcodes/hdf5/HDF5IO.h:21,
                 from /build/csound/src/csound-6.14.0/Opcodes/hdf5/HDF5IO.c:20:
/usr/include/H5Opublic.h:187:15: note: declared here
  187 | H5_DLL herr_t H5Oget_info3(hid_t loc_id, H5O_info2_t *oinfo, unsigned fields);
      |               ^~~~~~~~~~~~
make[2]: *** [Opcodes/hdf5/CMakeFiles/hdf5ops.dir/build.make:83: Opcodes/hdf5/CMakeFiles/hdf5ops.dir/HDF5IO.c.o] Error 1
make[2]: Leaving directory '/build/csound/src/csound-6.14.0/build'
make[1]: *** [CMakeFiles/Makefile2:1887: Opcodes/hdf5/CMakeFiles/hdf5ops.dir/all] Error 2
make[1]: Leaving directory '/build/csound/src/csound-6.14.0/build'
make: *** [Makefile:183: all] Error 2
make: Leaving directory '/build/csound/src/csound-6.14.0/build'
```

Full build log:
[csound-6.14-hdf5_build.log](https://github.com/csound/csound/files/4465288/csound-6.14-hdf5_build.log)

--- a/Opcodes/hdf5/CMakeLists.txt
+++ b/Opcodes/hdf5/CMakeLists.txt
@@ -4,6 +4,15 @@ if(BUILD_HDF5_OPCODES)
     find_package(HDF5)
 
     if(HDF5_FOUND)
+        # split version string, as cmake doesn't do this automatically
+        string(REPLACE "." ";" VERSION_LIST ${HDF5_VERSION})
+        list(GET VERSION_LIST 0 HDF5_VERSION_MAJOR)
+        list(GET VERSION_LIST 1 HDF5_VERSION_MINOR)
+        list(GET VERSION_LIST 2 HDF5_VERSION_PATCH)
+        # add definitions so we can use them in the preprocessor
+        ADD_DEFINITIONS(-DHDF5_VERSION_MAJOR=${HDF5_VERSION_MAJOR})
+        ADD_DEFINITIONS(-DHDF5_VERSION_MINOR=${HDF5_VERSION_MINOR})
+        ADD_DEFINITIONS(-DHDF5_VERSION_PATCH=${HDF5_VERSION_PATCH})
         if(MSVC)
             make_plugin(hdf5ops HDF5IO.c)
             target_include_directories(hdf5ops PRIVATE ${HDF5_INCLUDE_DIRS})

--- a/Opcodes/hdf5/HDF5IO.c
+++ b/Opcodes/hdf5/HDF5IO.c
@@ -229,7 +229,11 @@ void HDF5IO_readStringAttribute(CSOUND *csound, HDF5File *self,
 {
     hid_t dataSetID = H5Dopen2(self->fileHandle, datasetName, H5P_DEFAULT);
     H5O_info_t oinfo;
+#if HDF5_VERSION_MAJOR == 1 && HDF5_VERSION_MINOR >= 12
+    HDF5ERROR(H5Oget_info1(dataSetID, &oinfo));
+#else
     HDF5ERROR(H5Oget_info(dataSetID, &oinfo));
+#endif
     hid_t attributeID = H5Aopen_by_idx(dataSetID, ".", H5_INDEX_CRT_ORDER,
                                        H5_ITER_INC, 0, H5P_DEFAULT, H5P_DEFAULT);
     HDF5ERROR(attributeID);

